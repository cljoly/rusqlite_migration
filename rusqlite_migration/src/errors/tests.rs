use std::num::NonZeroUsize;

use super::*;

// We should be able to convert rusqlite errors transparently
#[test]
fn test_rusqlite_error_conversion() {
    assert!(matches!(
        Error::from(rusqlite::Error::MultipleStatement),
        Error::RusqliteError { query: _, err: _ }
    ));

    let hook_error = HookError::from(rusqlite::Error::MultipleStatement);
    assert!(matches!(&hook_error, &HookError::RusqliteError(_)));
    assert!(matches!(
        Error::from(hook_error),
        Error::RusqliteError { query: _, err: _ },
    ));
}

// Check that Unrecognized errors correctly implement PartialEq, namely that
// > a != b if and only if !(a == b).
// from https://doc.rust-lang.org/std/cmp/trait.PartialEq.html
#[test]
fn test_unrecognized_errors() {
    let u1 = Error::Unrecognized(Box::new(Error::Hook(String::new())));
    let u2 = Error::Unrecognized(Box::new(Error::Hook(String::new())));
    let u3 = Error::Unrecognized(Box::new(Error::Hook("1".to_owned())));
    let u4 = Error::Unrecognized(Box::new(Error::Hook(String::new())));
    let u5 = Error::FileLoad("1".to_owned());
    let u6 = Error::Unrecognized(Box::new(Error::Hook(String::new())));

    for (e1, e2) in &[(u1, u2), (u3, u4), (u5, u6)] {
        assert!(e1 != e2);
        assert!(!(e1 == e2));
    }
}

#[test]
// Errors on specified schema versions should be equal if and only if all versions are
// equal
fn test_specified_schema_version_error() {
    assert_eq!(
        Error::SpecifiedSchemaVersion(SchemaVersionError::TargetVersionOutOfRange {
            specified: SchemaVersion::Outside(NonZeroUsize::new(10).unwrap()),
            highest: SchemaVersion::Inside(NonZeroUsize::new(4).unwrap()),
        }),
        Error::SpecifiedSchemaVersion(SchemaVersionError::TargetVersionOutOfRange {
            specified: SchemaVersion::Outside(NonZeroUsize::new(10).unwrap()),
            highest: SchemaVersion::Inside(NonZeroUsize::new(4).unwrap()),
        }),
    );
    assert_ne!(
        Error::SpecifiedSchemaVersion(SchemaVersionError::TargetVersionOutOfRange {
            specified: SchemaVersion::Outside(NonZeroUsize::new(9).unwrap()),
            highest: SchemaVersion::Inside(NonZeroUsize::new(4).unwrap()),
        }),
        Error::SpecifiedSchemaVersion(SchemaVersionError::TargetVersionOutOfRange {
            specified: SchemaVersion::Outside(NonZeroUsize::new(10).unwrap()),
            highest: SchemaVersion::Inside(NonZeroUsize::new(4).unwrap()),
        }),
    );
}

// Two errors with different queries or errors should be considered different
#[test]
fn test_rusqlite_error_query() {
    assert_ne!(
        Error::RusqliteError {
            query: "SELECTTT".to_owned(),
            err: rusqlite::Error::InvalidQuery
        },
        Error::RusqliteError {
            query: "SSSELECT".to_owned(),
            err: rusqlite::Error::InvalidQuery
        }
    );
    assert_ne!(
        Error::RusqliteError {
            query: "SELECT".to_owned(),
            err: rusqlite::Error::MultipleStatement
        },
        Error::RusqliteError {
            query: "SELECT".to_owned(),
            err: rusqlite::Error::InvalidQuery
        }
    )
}

// Two errors with different file load errors should be considered different
#[test]
fn test_rusqlite_error_file_load() {
    assert_ne!(
        Error::FileLoad("s1".to_owned()),
        Error::FileLoad("s2".to_owned())
    )
}

// Two errors with different foreign key checks should be considered different
#[test]
fn test_rusqlite_error_fkc() {
    assert_ne!(
        Error::ForeignKeyCheck(vec![ForeignKeyCheckError {
            table: "t1".to_owned(),
            rowid: 1,
            parent: "t2".to_owned(),
            fkid: 3
        }]),
        Error::ForeignKeyCheck(vec![ForeignKeyCheckError {
            table: "t1".to_owned(),
            rowid: 3,
            parent: "t2".to_owned(),
            fkid: 3
        }]),
    )
}

// Hook error conversion preserves the message
#[test]
fn test_hook_conversion_msg() {
    let msg = String::from("some error encountered in the hook");
    let hook_error = HookError::Hook(msg.clone());

    assert_eq!(Error::from(hook_error), Error::Hook(msg))
}
