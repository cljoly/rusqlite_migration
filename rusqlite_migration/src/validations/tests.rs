// SPDX-License-Identifier: Apache-2.0
// Copyright Cl√©ment Joly and contributors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::*;
use crate::Migrations;

use crate::tests::helpers::{
    all_valid_down, m_valid0_up, m_valid11_up, m_valid20_up, m_valid_fk_up,
};

#[test]
fn test_empty_migrations() {
    let migrations = Migrations::from_slice(&[]);
    assert_eq!(Validations::upward().validate(&migrations), Ok(()));
    assert_eq!(Validations::everything().validate(&migrations), Ok(()));
}

#[test]
fn test_missing_downward_start() {
    let mut missing_start = all_valid_down();
    missing_start[0] = m_valid0_up();

    let migrations = Migrations::new(missing_start);
    insta::assert_debug_snapshot!(Validations::everything().validate(&migrations));

    // Additional checks for consistency
    assert_eq!(Validations::upward().validate(&migrations), Ok(()));
    {
        use std::error::Error;

        let v = Validations::everything().validate(&migrations).unwrap_err();
        insta::assert_snapshot!("error_missing_downward_pretty_print", v);
        assert!(v.source().is_none());
    }
}

#[test]
fn test_missing_downward_middle() {
    let mut missing_middle = all_valid_down();
    missing_middle[3] = m_valid20_up();

    let migrations = Migrations::new(missing_middle);
    insta::assert_debug_snapshot!(Validations::everything().validate(&migrations));
}

#[test]
fn test_missing_downward_end() {
    let mut missing_end = all_valid_down();
    let len = missing_end.len();
    missing_end[len - 1] = m_valid_fk_up();

    let migrations = Migrations::new(missing_end);
    insta::assert_debug_snapshot!(Validations::everything().validate(&migrations));
}

#[test]
fn test_missing_downward_multiple() {
    let mut missing_multiple = all_valid_down();
    let len = missing_multiple.len();
    missing_multiple[2] = m_valid11_up();
    missing_multiple[3] = m_valid20_up();
    missing_multiple[len - 1] = m_valid_fk_up();

    let migrations = Migrations::new(missing_multiple);
    insta::assert_debug_snapshot!(
        "multiple_missing_multiple_errors",
        Validations::everything().validate(&migrations)
    );
}

#[test]
fn test_invalid_down_migration() {
    use std::error::Error;

    let mut invalid_start = all_valid_down();
    invalid_start[0] = m_valid0_up().down("Invalid sql");

    let migrations = Migrations::new(invalid_start);
    let v = Validations::everything().validate(&migrations);

    insta::assert_debug_snapshot!("full_error", v);
    assert_eq!(
        Validations::upward()
            .require_downward()
            .validate(&migrations),
        v
    );
    assert_eq!(
        Validations::upward()
            .check_downward_if_present()
            .validate(&migrations),
        v
    );
    assert!(Validations::upward().validate(&migrations).is_ok());

    insta::assert_debug_snapshot!("source", v.unwrap_err().source());
}

#[test]
fn test_rusqlite_error_conversion() {
    let rusqlite_e = rusqlite::Error::InvalidQuery;
    let e = Error::from(rusqlite_e);
    assert!(matches!(
        e,
        Error::RusqliteMigration(crate::Error::RusqliteError { .. })
    ));
    insta::assert_snapshot!(e)
}
