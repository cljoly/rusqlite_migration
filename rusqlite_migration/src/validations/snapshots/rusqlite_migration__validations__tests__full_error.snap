---
source: rusqlite_migration/src/validations/tests.rs
expression: v
snapshot_kind: text
---
Err(
    RusqliteMigration(
        RusqliteError {
            query: "Invalid sql",
            err: SqliteFailure(
                Error {
                    code: Unknown,
                    extended_code: 1,
                },
                Some(
                    "near \"Invalid\": syntax error",
                ),
            ),
        },
    ),
)
